src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1724,104 +1724,300 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, Diagnost
         /// </summary>
         private BoundExpression BindDeconstructionAssignment(AssignmentExpressionSyntax node, DiagnosticBag diagnostics)
         {
-            SeparatedSyntaxList<ArgumentSyntax> arguments = ((TupleExpressionSyntax)node.Left).Arguments;
-
-            // receiver for Deconstruct
+            // receiver for first Deconstruct step
             var boundRHS = BindValue(node.Right, diagnostics, BindValueKind.RValue);
 
-            int numElements = arguments.Count;
-            Debug.Assert(numElements >= 2); // this should not have parsed as a tuple.
+            SeparatedSyntaxList<ArgumentSyntax> arguments = ((TupleExpressionSyntax)node.Left).Arguments;
+            ImmutableArray<BoundExpression> checkedVariables = BindDeconstructionVariables(arguments, diagnostics);
 
-            // bind the variables and check they can be assigned to
-            var checkedVariablesBuilder = ArrayBuilder<BoundExpression>.GetInstance(numElements);
-            foreach (var argument in arguments)
+            var deconstructionSteps = ArrayBuilder<BoundDeconstructionAssignmentOperator.DeconstructStep>.GetInstance(1);
+            var assignmentSteps = ArrayBuilder<BoundDeconstructionAssignmentOperator.AssignmentStep>.GetInstance(1);
+            try
             {
-                var boundVariable = BindExpression(argument.Expression, diagnostics, invoked: false, indexed: false);
-                var checkedVariable = CheckValue(boundVariable, BindValueKind.Assignment, diagnostics);
+                if ((object)boundRHS.Type == null)
+                {
+                    if (boundRHS.Kind == BoundKind.TupleLiteral)
+                    {
+                        // tuple literal without type such as `(null, null)`, let's fix it up by peeking at the LHS
+                        TypeSymbol lhsAsTuple = MakeTupleTypeFromDeconstructionLHS(checkedVariables, diagnostics);
+                        boundRHS = GenerateConversionForAssignment(lhsAsTuple, boundRHS, diagnostics);
+                    }
+                    else
+                    {
+                        // expression without type such as `null`
+                        Error(diagnostics, ErrorCode.ERR_DeconstructRequiresExpression, node);
+                        return BadExpression(node, checkedVariables.Concat(boundRHS).ToArray());
+                    }
+                }
 
-                checkedVariablesBuilder.Add(checkedVariable);
-            }
+                if (!DeconstructIntoSteps(new BoundDeconstructValuePlaceholder(node.Right, boundRHS.Type), node, diagnostics, checkedVariables, deconstructionSteps, assignmentSteps))
+                {
+                    return new BoundDeconstructionAssignmentOperator(node, checkedVariables, boundRHS, OneOrMany.Create(ImmutableArray<BoundDeconstructionAssignmentOperator.DeconstructStep>.Empty),
```

Comment:
The `OneOrMany<T>` argument is neither one nor many. Is the `OneOrMany` type misnamed or should the argument in this case be simply `ImmutableArray<T>`?